File: Day07/day7.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.*;

public class day7 {
    public static void main(String[] args) {
        Directory headDir = new Directory();
        headDir.name = "/";
        Directory currentDir = headDir;
        try {
            BufferedReader reader = new BufferedReader(new FileReader("input.txt"));
            String input = reader.readLine();
            while ((input = reader.readLine()) != null) {
                String[] cmd = input.split(" ");
                if (cmd[0].equals("$")) {
                    if (cmd[1].equals("cd")) {
                        if (cmd[2].equals("..")) {
                            currentDir = currentDir.parent;
                            continue;
                        }
                        for (Directory dir : currentDir.children) {
                            if (dir.name.equals(cmd[2])) {
                                currentDir = dir;
                                break;
                            }
                        }
                    }
                } else if (cmd[0].equals("dir")) {
                    Directory childDir = new Directory();
                    childDir.parent = currentDir;
                    childDir.name = cmd[1];
                    currentDir.children.add(childDir);
                } else {
                    currentDir.fileSizes += Integer.parseInt(cmd[0]);
                }
            }
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Part 1 Answer
        System.out.println(partOneAnswer(headDir));

        // Part 2 Answer
        System.out.println(partTwoAnswer(headDir, 30000000 - (70000000 - countSize(headDir))));
    }

    public static int countSize(Directory directory) {
        int totalSize = directory.fileSizes;
        for (Directory dir : directory.children) {
            totalSize += countSize(dir);
        }
        return totalSize;
    }

    public static int partOneAnswer(Directory directory) {
        int result = 0;
        int totalSize = countSize(directory);

        if (totalSize <= 100000) {
            result += totalSize;
        }

        for (Directory dir : directory.children) {
            result += partOneAnswer(dir);
        }

        return result;
    }

    public static int partTwoAnswer(Directory directory, int spaceNeeded) {
        int resultSize = countSize(directory);
        if (resultSize < spaceNeeded) {
            return Integer.MAX_VALUE;
        }
        for (Directory dir : directory.children) {
            int resultSubSize = partTwoAnswer(dir, spaceNeeded);
            if (resultSubSize < resultSize) {
                resultSize = resultSubSize;
            }
        }
        return resultSize;
    }
}

class Directory {
    String name;
    Directory parent;
    ArrayList<Directory> children = new ArrayList<>();
    int fileSizes = 0;
}

File: Day07/day7.py
def check(dir):
    if dir in memo:
        return memo[dir]
    if dir in storage:
        result = storage[dir]
    else:
        result = 0
    if dir in track:
        for direc in track[dir]:
            if direc in memo:
                result += memo[direc]
            else:
                result += check(direc)
    if dir not in memo:
        memo[dir] = result
    return result

memo = {}
all_dir = []
track = {}
storage = {}
curr = []
with open("input.txt") as f:
    for i in f:
        i = i.strip("\n").split(" ")

        if ' '.join(curr) not in all_dir:
            all_dir.append(' '.join(curr))

        if i[0] == "$":
            if i[1] == "cd":
                if i[2] == "..":
                    curr.pop()
                else:
                    curr.append(i[2])
        elif i[0] == "dir":
            temp = curr.copy()
            temp.append(i[1])
            if ' '.join(temp) not in all_dir:
                all_dir.append(' '.join(temp))
            if ' '.join(curr) not in track:
                track[' '.join(curr)] = [' '.join(temp)]
            else:
                if temp not in track[' '.join(curr)]:
                    track[' '.join(curr)].append(' '.join(temp))
        elif i[0] == "ls":
            continue
        else:
            size = int(i[0])
            if ' '.join(curr) not in storage:
                storage[' '.join(curr)] = size
            else:
                storage[' '.join(curr)] += size
part_one_answer = 0
for dir in all_dir:
    sizing = check(dir)
    if sizing <= 100000:
        part_one_answer += sizing

# Part 1 Answer
print(part_one_answer)

space_to_delete = memo["/"] - 40000000
all_used_spaces = list(memo.values())
all_used_spaces.sort()
for used_space in all_used_spaces:
    if space_to_delete < used_space:

        # Part 2 Answer
        print(used_space)
        break

